src/events/backup.py

```diff
@@ -139,6 +149,128 @@ def _on_list_backups_action(self, event: ActionEvent):
+        unit_state = self.charm.state.unit_server
+
+        if not cluster_state.id_to_restore:
+            if unit_state.restore_progress is not RestoreStep.NOT_STARTED:
+                self.charm.state.unit_server.update(
+                    {"restore-progress": RestoreStep.NOT_STARTED.value}
+                )
+                self.charm._set_status(self.charm.state.ready)
+            return
+
+        if self.charm.unit.is_leader():
+            self._maybe_progress_step()
+
+        match cluster_state.restore_instruction, unit_state.restore_progress:
+            case RestoreStep.STOP_WORKFLOW, RestoreStep.NOT_STARTED:
+                self._stop_workflow()
+            case RestoreStep.RESTORE, RestoreStep.STOP_WORKFLOW:
+                self._download_and_restore()
+            case RestoreStep.RESTART, RestoreStep.RESTORE:
+                self._restart_workflow()
+            case RestoreStep.CLEAN, RestoreStep.RESTART:
+                self._cleaning()
+            case _:
+                pass
+
+    def _maybe_progress_step(self):
+        """Check that all units are done with the current instruction and move to the next if applicable."""
+        current_instruction = self.charm.state.cluster.restore_instruction
+        next_instruction = current_instruction.next_step()
+
+        if all(
+            (unit.restore_progress is current_instruction for unit in self.charm.state.servers)
+        ):
+            payload = {"restore-instruction": next_instruction.value}
+            if current_instruction is RestoreStep.CLEAN:
+                payload = payload | {"id-to-restore": "", "to_restore": ""}
+                # Update ACLs for already related clients and trigger a relation-changed
+                # on their side to enable them to reconnect.
+                self.charm.update_client_data()
+                self.charm.quorum_manager.update_acls()
+
+            self.charm.state.cluster.update(payload)
+
+    def _stop_workflow(self) -> None:
+        self.charm._set_status(Status.ONGOING_RESTORE)
+        logger.info("Restoring - stopping workflow")
+        self.charm.workload.stop()
+        self.charm.state.unit_server.update({"restore-progress": RestoreStep.STOP_WORKFLOW.value})
+
+    def _download_and_restore(self) -> None:
+        logger.info("Restoring - restore snapshot")
+        self.backup_manager.restore_snapshot(
+            self.charm.state.cluster.id_to_restore, self.charm.workload
+        )
+        self.charm.state.unit_server.update({"restore-progress": RestoreStep.RESTORE.value})
+
+    def _restart_workflow(self) -> None:
+        logger.info("Restoring - restarting workflow")
+        self.charm.workload.restart()
+        self.charm.state.unit_server.update({"restore-progress": RestoreStep.RESTART.value})
+
+    @retry(
+        wait=wait_fixed(5),
+        stop=stop_after_attempt(3),
+        retry=retry_if_result(lambda res: res is False),
+    )
```

Comment:
So, the idea is to give some leeway for the units to start and sync, but also making sure they are running. We cannot include this in the previous step because the execution model is running one unit at a time, so if we retry on the first unit, the other ones will wait for it to finish or error before beginning the restart.
After 3 unsuccessful restarts, the event will be in error because of `RetryError`, requiring a manual intervention to address the issue

I did not have any issue with the process before adding this retry though, it might just be me overthinking things.

Reply:
Ah it's probably fine, keep it 👍🏾